File: BoxOfDice/GameEngine.swift
//
//  GameEngine.swift
//  BoxOfDice
//

import Foundation

// Platform-independent core game logic shared by iOS today and a future watchOS target.
// Keep SwiftUI, UIKit, haptics, sounds, persistence, and device-specific layout out of this file.
// A future WatchGameView can drive this engine directly or through GameViewModel.
struct Tile: Identifiable, Equatable {
    let id: Int
    var isOpen: Bool = true
}

struct MoveRecord: Identifiable {
    let id = UUID()
    let dice: [Int]
    let closedTiles: [Int]
    let scoreAfterMove: Int

    var diceTotal: Int { dice.reduce(0, +) }
}

enum GameState: Equatable {
    case waitingToRoll
    case selecting
    case gameOver(won: Bool)
}

struct SeededRandomNumberGenerator: RandomNumberGenerator {
    private var state: UInt64

    init(seed: UInt64) {
        state = seed == 0 ? 0x9E3779B97F4A7C15 : seed
    }

    mutating func next() -> UInt64 {
        state &+= 0x9E3779B97F4A7C15
        var value = state
        value = (value ^ (value >> 30)) &* 0xBF58476D1CE4E5B9
        value = (value ^ (value >> 27)) &* 0x94D049BB133111EB
        return value ^ (value >> 31)
    }
}

struct GameEngine {

    // MARK: - State

    private(set) var settings: GameSettings
    private(set) var tiles: [Tile]
    // Dice stored as array of up to 3 values; 0 = inactive die slot.
    private(set) var dice: [Int] = [0, 0, 0]
    private(set) var selectedTiles: Set<Int> = []
    private(set) var gameState: GameState = .waitingToRoll
    private(set) var moveHistory: [MoveRecord] = []
    private var randomGenerator: SeededRandomNumberGenerator?

    init(settings: GameSettings = .default, seed: UInt64? = nil) {
        self.settings = settings
        self.tiles = (1...settings.tileCount).map { Tile(id: $0) }
        if let seed {
            self.randomGenerator = SeededRandomNumberGenerator(seed: seed)
        }
    }

    // MARK: - Computed Properties

    var diceTotal: Int { dice.reduce(0, +) }

    var score: Int {
        tiles.filter { $0.isOpen }.reduce(0) { $0 + $1.id }
    }

    var selectionTotal: Int {
        selectedTiles.reduce(0, +)
    }

    var turnCount: Int { moveHistory.count }

    var canConfirm: Bool {
        !selectedTiles.isEmpty && isSelectionAllowed(selectedTiles) && selectionTotal == diceTotal
    }

    var canUndo: Bool {
        gameState == .waitingToRoll && !moveHistory.isEmpty
    }

    var hasRolled: Bool { dice.contains(where: { $0 > 0 }) }

    var openTiles: [Int] {
        tiles.filter { $0.isOpen }.map { $0.id }
    }

    var currentDieCount: Int {
        switch settings.diceMode {
        case .alwaysTwoDice:
            return settings.baseDiceCount
        case .oneDieWhenLowTilesRemain:
            return openTiles.allSatisfy { $0 <= 6 } ? 1 : settings.baseDiceCount
        }
    }

    var isGameOver: Bool {
        if case .gameOver = gameState { return true }
        return false
    }

    var isWinner: Bool {
        if case .gameOver(let won) = gameState { return won }
        return false
    }

    // MARK: - Actions

    /// Rolls the dice, clears any pending selection, then transitions to .selecting
    /// or .gameOver depending on whether a valid move exists.
    mutating func roll(dice rolledDice: [Int]? = nil) {
        guard case .waitingToRoll = gameState else { return }
        selectedTiles = []
        dice = rolledDice ?? randomDice()
        gameState = hasValidMoves() ? .selecting : .gameOver(won: false)
    }

    /// Toggles a tile in or out of the pending selection.
    mutating func toggleTile(_ number: Int) {
        guard case .selecting = gameState else { return }
        guard tiles.first(where: { $0.id == number })?.isOpen == true else { return }
        if selectedTiles.contains(number) {
            selectedTiles.remove(number)
        } else {
            selectedTiles.insert(number)
        }
    }

    /// Closes the selected tiles when they sum to the dice total, then transitions
    /// to .gameOver(won: true) if all tiles are closed, otherwise back to .waitingToRoll.
    mutating func confirmSelection() {
        guard canConfirm, case .selecting = gameState else { return }
        let closedTiles = selectedTiles.sorted()
        for i in tiles.indices where selectedTiles.contains(tiles[i].id) {
            tiles[i].isOpen = false
        }
        selectedTiles = []
        let won = tiles.allSatisfy { !$0.isOpen }
        moveHistory.append(MoveRecord(dice: dice, closedTiles: closedTiles, scoreAfterMove: score))
        gameState = won ? .gameOver(won: true) : .waitingToRoll
    }

    mutating func undoLastMove() {
        guard canUndo, let lastMove = moveHistory.popLast() else { return }
        for i in tiles.indices where lastMove.closedTiles.contains(tiles[i].id) {
            tiles[i].isOpen = true
        }
        dice = [0, 0, 0]
        selectedTiles = []
        gameState = .waitingToRoll
    }

    /// Resets all state to the beginning of a new game.
    mutating func reset(settings newSettings: GameSettings? = nil, seed: UInt64? = nil) {
        if let newSettings {
            settings = newSettings
        }
        if let seed {
            randomGenerator = SeededRandomNumberGenerator(seed: seed)
        } else {
            randomGenerator = nil
        }
        tiles = (1...settings.tileCount).map { Tile(id: $0) }
        dice = [0, 0, 0]
        selectedTiles = []
        moveHistory = []
        gameState = .waitingToRoll
    }

    mutating func randomDice() -> [Int] {
        let count = currentDieCount
        var result = [0, 0, 0]
        for i in 0..<count {
            result[i] = randomDie()
        }
        return result
    }

    // MARK: - Move Validation

    /// Returns true if any allowed selection of currently open tiles sums to the dice total.
    func hasValidMoves() -> Bool {
        !validMoves().isEmpty
    }

    func bestHint() -> [Int]? {
        validMoves().min { lhs, rhs in
            if lhs.count == rhs.count { return lhs.lexicographicallyPrecedes(rhs) }
            return lhs.count < rhs.count
        }
    }

    func validMoves() -> [[Int]] {
        allowedSubsets(openTiles, target: diceTotal)
    }

    // MARK: - Private

    private mutating func randomDie() -> Int {
        if var generator = randomGenerator {
            let value = Int.random(in: 1...6, using: &generator)
            randomGenerator = generator
            return value
        }
        return Int.random(in: 1...6)
    }

    private func isSelectionAllowed(_ selection: Set<Int>) -> Bool {
        switch settings.moveRule {
        case .anyCombination:
            return true
        case .oneOrTwoTiles:
            return selection.count <= 2
        }
    }

    private func allowedSubsets(_ numbers: [Int], target: Int) -> [[Int]] {
        guard target > 0 else { return [] }
        let n = numbers.count
        var matches: [[Int]] = []

        for mask in 1..<(1 << n) {
            var subset: [Int] = []
            var sum = 0
            for i in 0..<n where mask & (1 << i) != 0 {
                sum += numbers[i]
                subset.append(numbers[i])
                if sum > target { break }
            }
            if sum == target && isSelectionAllowed(Set(subset)) {
                matches.append(subset)
            }
        }

        return matches
    }
}

File: app/src/main/java/com/example/boxofdice/engine/GameEngine.kt
package com.example.boxofdice.engine

import com.example.boxofdice.model.DiceMode
import com.example.boxofdice.model.GameMode
import com.example.boxofdice.model.MoveRule

/**
 * Self-contained game engine for Box of Dice (Shut the Box).
 *
 * @param currentGameMode  Mode controlling tile count and base dice count.
 * @param diceRoller       Injectable dice source — receives the number of dice to roll and
 *                         returns a list of that size. Defaults to (1..6).random() per die.
 *                         Override in tests to make rolls deterministic.
 * @param initialOpenTiles Starting set of open tile numbers. Defaults to every tile for the mode.
 *                         Supply a subset in tests to fast-forward to a specific board state.
 * @param diceMode         Whether all dice are always thrown, or a single die once only tiles
 *                         1–6 remain open.
 * @param moveRule         Whether any combination of open tiles may be closed, or only one/two.
 */
class GameEngine(
    val currentGameMode: GameMode,
    private val diceRoller: (count: Int) -> List<Int> = { n -> List(n) { (1..6).random() } },
    initialOpenTiles: Set<Int> = (1..currentGameMode.tileCount).toSet(),
    val diceMode: DiceMode = DiceMode.ALWAYS_ALL,
    val moveRule: MoveRule = MoveRule.ANY_COMBINATION
) {
    /** A single confirmed move, kept so it can be undone. */
    data class MoveRecord(val dice: List<Int>, val closedTiles: Set<Int>)

    // ── Internal mutable state
    // ────────────────────────────────────────────────

    private val _open: MutableSet<Int> =
        initialOpenTiles.toMutableSet()commit

    private val _closed: MutableSet<Int> =
        ((1..currentGameMode.tileCount).toSet() - initialOpenTiles).toMutableSet()

    private val _selected: MutableSet<Int> =
        mutableSetOf()

    private var _diceValues: List<Int> =
        emptyList()

    private var _gameOver: Boolean =
        false

    private val _history: MutableList<MoveRecord> =
        mutableListOf()

    // ── Public read-only state ────────────────────────────────────────────────

    /** Numbers of tiles that have not yet been closed. */
    val openTiles: Set<Int> get() = _open.toSet()

    /** Numbers of tiles that have been confirmed closed this game. */
    val closedTiles: Set<Int> get() = _closed.toSet()

    /** Subset of [openTiles] the player has highlighted for the current move. */
    val selectedTiles: Set<Int> get() = _selected.toSet()

    /** Values shown on the dice after the most recent [rollDice], empty when awaiting a roll. */
    val diceValues: List<Int> get() = _diceValues.toList()

    /** Sum of the current dice — the target the player must match with tile selections. */
    val currentTargetSum: Int get() = _diceValues.sum()

    /** Confirmed moves this game, oldest first. */
    val moveHistory: List<MoveRecord> get() = _history.toList()

    /** Number of confirmed moves so far. */
    val turnCount: Int get() = _history.size

    /**
     * How many dice the next roll will throw. With [DiceMode.ONE_DIE_WHEN_LOW] this drops to a
     * single die once every open tile is 6 or lower; otherwise it is the mode's base dice count.
     */
    val currentDieCount: Int
        get() = when (diceMode) {
            DiceMode.ALWAYS_ALL -> currentGameMode.diceCount
            DiceMode.ONE_DIE_WHEN_LOW ->
                if (_open.isNotEmpty() && _open.all { it <= 6 }) 1 else currentGameMode.diceCount
        }

    // ── Actions ───────────────────────────────────────────────────────────────

    /**
     * Roll the dice for this turn.
     * Clears any pending tile selection, generates [currentDieCount] dice via [diceRoller],
     * then immediately checks whether any valid move exists.
     * If none exists the game is over.
     */
    fun rollDice() {
        check(!_gameOver) { "Game is already over." }
        _selected.clear()
        _diceValues = diceRoller(currentDieCount)
        if (!hasValidMove()) {
            _gameOver = true
        }
    }

    /**
     * Toggle selection of an open tile.
     * Selecting a tile adds it to the pending move; selecting it again removes it.
     *
     * @throws IllegalStateException if dice have not been rolled yet, the game is over,
     *         or [tile] is not in [openTiles].
     */
    fun toggleTileSelection(tile: Int) {
        check(!_gameOver) { "Game is already over." }
        check(_diceValues.isNotEmpty()) { "Roll dice before selecting tiles." }
        check(tile in _open) { "Tile $tile is not open (open=$_open)." }
        if (tile in _selected) _selected.remove(tile) else _selected.add(tile)
    }

    /**
     * Returns true when the current selection is non-empty, allowed by the [moveRule], and its
     * sum equals [currentTargetSum]. This is the gate for [confirmMove].
     */
    fun canConfirmMove(): Boolean =
        _diceValues.isNotEmpty() &&
        _selected.isNotEmpty() &&
        isSelectionAllowed(_selected) &&
        _selected.sumOf { it } == currentTargetSum

    /**
     * Close all selected tiles and prepare for the next roll.
     * If the board is cleared the game ends with a perfect result.
     *
     * @throws IllegalStateException if [canConfirmMove] returns false.
     */
    fun confirmMove() {
        check(canConfirmMove()) {
            "Cannot confirm: selected ${_selected.sumOf { it }} ≠ target $currentTargetSum."
        }
        _history.add(MoveRecord(dice = _diceValues.toList(), closedTiles = _selected.toSet()))
        _open.removeAll(_selected)
        _closed.addAll(_selected)
        _selected.clear()
        _diceValues = emptyList()
        if (_open.isEmpty()) _gameOver = true
    }

    /** True when the most recent confirmed move can be undone (between rolls, game not over). */
    fun canUndo(): Boolean =
        !_gameOver && _diceValues.isEmpty() && _history.isNotEmpty()

    /**
     * Reopen the tiles closed by the most recent confirmed move and return to the
     * waiting-to-roll state. No-op when [canUndo] is false.
     */
    fun undoLastMove() {
        if (!canUndo()) return
        val last = _history.removeAt(_history.lastIndex)
        _open.addAll(last.closedTiles)
        _closed.removeAll(last.closedTiles)
        _selected.clear()
        _diceValues = emptyList()
    }

    /**
     * Returns true if any allowed subset of [openTiles] sums to [currentTargetSum].
     * With [MoveRule.ANY_COMBINATION] this uses 0/1 knapsack DP — O(n × target). With
     * [MoveRule.ONE_OR_TWO_TILES] only one- and two-tile selections count.
     * Always false before dice are rolled.
     */
    fun hasValidMove(): Boolean {
        if (_diceValues.isEmpty()) return false
        val target = currentTargetSum
        return when (moveRule) {
            MoveRule.ANY_COMBINATION -> hasSubsetWithSum(_open.toList(), target)
            MoveRule.ONE_OR_TWO_TILES -> hasOneOrTwoTileMove(_open.toList(), target)
        }
    }

    /**
     * All allowed tile selections that sum to the current dice total. Empty before a roll or when
     * the game is over. Honours the [moveRule].
     */
    fun validMoves(): List<List<Int>> {
        if (_diceValues.isEmpty() || _gameOver) return emptyList()
        return allowedSubsets(_open.toList().sorted(), currentTargetSum)
    }

    /**
     * The "best" hint: the valid move that closes the fewest tiles, breaking ties by the
     * lexicographically smallest set. Null when no move exists.
     */
    fun bestHint(): List<Int>? =
        validMoves().minWithOrNull(
            compareBy<List<Int>> { it.size }.thenComparator { a, b -> lexCompare(a, b) }
        )

    /** Sum of remaining open tile numbers — the player's current score (lower is better). */
    fun calculateScore(): Int = _open.sum()

    /** True when every tile has been closed (perfect game). */
    fun isBoardCleared(): Boolean = _open.isEmpty()

    /**
     * True when the game has ended — either because [rollDice] found no valid move,
     * or because [confirmMove] cleared the board.
     */
    fun isGameOver(): Boolean = _gameOver

    // ── Private helpers ───────────────────────────────────────────────────────

    private fun isSelectionAllowed(selection: Set<Int>): Boolean = when (moveRule) {
        MoveRule.ANY_COMBINATION -> true
        MoveRule.ONE_OR_TWO_TILES -> selection.size <= 2
    }

    private fun hasSubsetWithSum(numbers: List<Int>, target: Int): Boolean {
        if (target <= 0 || numbers.isEmpty()) return false
        if (target > numbers.sum()) return false      // quick-reject
        val dp = BooleanArray(target + 1)
        dp[0] = true
        for (num in numbers) {
            for (s in target downTo num) {
                if (dp[s - num]) dp[s] = true
            }
        }
        return dp[target]
    }

    private fun hasOneOrTwoTileMove(numbers: List<Int>, target: Int): Boolean {
        if (target <= 0) return false
        if (target in numbers) return true
        for (i in numbers.indices) {
            for (j in i + 1 until numbers.size) {
                if (numbers[i] + numbers[j] == target) return true
            }
        }
        return false
    }

    /** All subsets of [numbers] that sum to [target] and are permitted by the [moveRule]. */
    private fun allowedSubsets(numbers: List<Int>, target: Int): List<List<Int>> {
        if (target <= 0) return emptyList()
        val n = numbers.size
        val matches = mutableListOf<List<Int>>()
        // n is small (≤ 18); a bitmask sweep is clear and fast enough.
        for (mask in 1 until (1 shl n)) {
            var sum = 0
            val subset = mutableListOf<Int>()
            for (i in 0 until n) {
                if (mask and (1 shl i) != 0) {
                    sum += numbers[i]
                    subset.add(numbers[i])
                    if (sum > target) break
                }
            }
            if (sum == target && isSelectionAllowed(subset.toSet())) {
                matches.add(subset)
            }
        }
        return matches
    }

    private fun lexCompare(a: List<Int>, b: List<Int>): Int {
        val min = minOf(a.size, b.size)
        for (i in 0 until min) {
            val c = a[i].compareTo(b[i])
            if (c != 0) return c
        }
        return a.size.compareTo(b.size)
    }
}
